ambry-network/src/main/java/com.github.ambry.network/LocalNetworkClient.java

```diff
@@ -0,0 +1,116 @@
+ */
+public class LocalNetworkClient implements NetworkClient {
+  private final LocalRequestResponseChannel channel;
+  private final NetworkMetrics networkMetrics;
+  private final Time time;
+  private boolean closed = false;
+  private final int processorId;
+  private static final Logger logger = LoggerFactory.getLogger(LocalNetworkClient.class);
+
+  /**
+   * Instantiates a LocalNetworkClient.
+   * @param channel the {@link RequestResponseChannel} to send requests to.
+   * @param processorId the id used to isolate requests and responses sent across the shared channel.
+   * @param networkMetrics the metrics to track the network related metrics
+   * @param time The Time instance to use.
+   */
+  public LocalNetworkClient(LocalRequestResponseChannel channel, int processorId, NetworkMetrics networkMetrics,
+      Time time) {
+    this.channel = channel;
+    this.processorId = processorId;
+    this.networkMetrics = networkMetrics;
+    this.time = time;
+  }
+
+  /**
+   * Attempt to send the given requests and poll for responses from the network via the associated channel. Any
+   * requests that could not be sent out will be added to a queue. Every time this method is called, it will first
+   * attempt sending the requests in the queue (or time them out) and then attempt sending the newly added requests.
+   * @param requestInfos the list of {@link RequestInfo} representing the requests that need to be sent out. This
+   *                     could be empty.
+   * @param requestsToDrop the list of correlation IDs representing the requests that can be dropped by
+   *                       closing the connection.
+   * @param pollTimeoutMs the poll timeout.
+   * @return a list of {@link ResponseInfo} representing the responses received for any requests that were sent out
+   * so far.
+   * @throws IllegalStateException if the SocketNetworkClient is closed.
+   */
+  public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestInfos, Set<Integer> requestsToDrop,
+      int pollTimeoutMs) {
+    if (closed) {
+      throw new IllegalStateException("The client is closed.");
+    }
+    if (processorId < 0) {
+      throw new IllegalStateException("The processorId has not been set.");
+    }
+
+    // TODO: do anything with requestsToDrop?
+    // The AmbryRequest sessions run in a thread pool, and each thread knows when the response is back so we can poll it
+    long startTime = time.milliseconds();
+    try {
+      for (RequestInfo requestInfo : requestInfos) {
+        ByteBuffer buffer = LocalRequestResponseChannel.byteBufferFromPayload(requestInfo.getRequest());
+        channel.sendRequest(new LocalChannelRequest(requestInfo, processorId, new ByteBufferInputStream(buffer)));
+      }
+    } catch (Exception e) {
+      logger.error("Received an unexpected error during sendAndPoll(): ", e);
+      networkMetrics.networkClientException.inc();
+    }
+
+    List<ResponseInfo> responses = channel.receiveResponses(processorId);
```

Comment:
you might want to introduce some form of a timed wait mechanism here. In low qps situations this could run as a very tight loop when no responses are received